hpx/lcos/local/packaged_continuation.hpp

```diff
@@ -32,8 +32,49 @@
 #include <utility>
 
 ///////////////////////////////////////////////////////////////////////////////
+// forward declare the type we will get function annotations from
 namespace hpx { namespace lcos { namespace detail
 {
+    template <typename Continuation>
+    struct continuation_finalization;
+}}}
+
+namespace hpx { namespace traits {
+    #if defined(HPX_HAVE_THREAD_DESCRIPTION)
+        ///////////////////////////////////////////////////////////////////////////
+        // traits specialization to get annotation from dataflow_finalization
+        template <typename Continuation>
+        struct get_function_annotation<lcos::detail::continuation_finalization<Continuation>>
+        {
+            using function_type = typename Continuation::function_type;
+            static char const* call(
+                lcos::detail::continuation_finalization<Continuation> const& f) noexcept
+            {
+                char const* annotation = hpx::traits::get_function_annotation<
+                    typename hpx::util::decay<function_type>::type>::call(f.this_->f_);
+                return annotation;
+            }
+        };
+    #endif
+}}
```

Comment:
Is this specialization really needed? The code below already calls `get_function_annotation` on the embedded function `f_`. Please explain.

Reply:
The invocations for version with and without executors and different policies go through different places. I'm not sure I know which bits you are referring to as unneeded.

Reply:
Well, I think the specialization I attached this comment onto is not really needed... What is the use case (i.e. test) that is fixed by _this_ particular change?

Reply:
```
    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            NP_executor,
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation with executor")));
    }
```

Reply:
That code seems to work completely without your patch. I.e.
```
#include <hpx/hpx.hpp>
#include <hpx/hpx_main.hpp>

int main()
{
    hpx::future<int> f1 = hpx::make_ready_future(5);
     hpx::future<void> f2 = f1.then(hpx::parallel::execution::parallel_executor{},
        hpx::util::annotated_function(
            [](auto&& f1)
            {
                std::cout << hpx::threads::get_thread_description(
                    hpx::threads::get_self_id()) << "\n";
            }, "Continuation with executor"));
    f2.get();
    return 0;
}
```
Will print `Continuation with executor` without your patch. What do I miss?

BTW: could you add some tests for your changes, please?

Reply:
try it with 
```
    hpx::threads::scheduled_executor NP_executor =
           hpx::threads::executors::pool_executor(
               "default", hpx::threads::thread_priority_default);
  
```

Reply:
There isn't really a way of testing this. you have to use tools like vampir to load the generated otf files to see if the functions are visible. The file `tests/unit/resource/scheduler_priority_check.cpp` on my current branch has been extended with annotations for some of these and I will commit it to the branch once I'm happy that apex works. The last time I was able to use apex was back in early 2018, so there's no great rush with this.

Reply:
The tests I'm running are ...

```
    hpx::threads::scheduled_executor NP_executor =
        hpx::threads::executors::pool_executor(
            "default", hpx::threads::thread_priority_default);

    std::vector<hpx::future<void>> results;

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        hpx::future<int> f2 = hpx::make_ready_future(5);
        results.emplace_back(hpx::dataflow(
            NP_executor,
            hpx::util::annotated_function(
                [](auto &&f1, auto &&f2) {
                    dummy_task(std::size_t(1000));
                },
                "Dataflow with executor"), f1, f2));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        hpx::future<int> f2 = hpx::make_ready_future(5);
        results.emplace_back(hpx::dataflow(
            hpx::util::annotated_function(
                [](auto &&f1, auto &&f2) {
                    dummy_task(std::size_t(1000));
                },
                "Dataflow without executor"), f1, f2));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        hpx::future<int> f2 = hpx::make_ready_future(5);
        results.emplace_back(hpx::dataflow(
            hpx::launch::sync,
            hpx::util::annotated_function(
                [](auto &&f1, auto &&f2) {
                    dummy_task(std::size_t(1000));
                },
                "Dataflow with sync policy"), f1, f2));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            NP_executor,
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation with executor")));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation without executor")));
    }

    {
        hpx::future<int> f1 = hpx::make_ready_future(5);
        results.emplace_back(f1.then(
            hpx::launch::sync,
            hpx::util::annotated_function(
                [](auto &&f1) {
                    dummy_task(std::size_t(1000));
                },
                "Continuation with sync policy")));
    }

    // wait for completion
    hpx::when_all(results).get();
```

Reply:
I should add some other policies too ...

Reply:
I forgot to add the combinations that use `hpx:unwrapping` - I will commit a test shortly to the branch

Reply:
I still believe, that this specialization is not really needed. Please verify. The code below already calls `get_function_annotation` on the embedded function `f_`. That should already be sufficient.